### PR TITLE
feat: health status implementation

### PR DESCRIPTION
**Pull request by kohlisid**
Health status definitions 

It is divided into two parts:
1. Pipeline Resource Health: It is based on the health of each vertex in the pipeline
2. Data Criticality: It is based on the data movement of the pipeline

**Resource Health** can be "healthy (0) | unhealthy (1) | paused (3) | unknown (4)".

Resource Health purely means it is up and running.
Resource health will be the max(health) based of each vertex's health

Resource health checks if all the pods are in running state for the pipeline, and also for paused, unknown pipeline etc

**Data Criticality** on the other end shows whether the pipeline is working as expected.
It represents the pending messages, lags, etc.
Data Criticality can be "ok (0) | warning (1) | critical (2)".
A backlogged pipeline can be healthy even though it has an increasing back-pressure.

For data criticality the timeline data is populated for the pipeline, and if the average usage lies above the thresholds then the required state is assigned. For critical states we have the option do a lookback and only assign it to be critical if we see a predefined number of critical state in a lookback period window.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -5,6 +5,7 @@ go 1.20
 require (
 	github.com/IBM/sarama v1.40.1
 	github.com/Masterminds/sprig/v3 v3.2.2
+	github.com/VividCortex/ewma v1.2.0
```
Comment (vigith) [inline go.mod:8]: can we do it ourselves? this is quite simple to do, right? WDYT?

File: pkg/daemon/server/daemon_server.go [hunk 2]
```diff
@@ -138,11 +138,23 @@ func (ds *daemonServer) Run(ctx context.Context) error {
 	go func() { _ = httpServer.Serve(httpL) }()
 	go func() { _ = tcpm.Serve() }()
 
+	pipelineMetadataQuery, err := service.NewPipelineMetadataQuery(isbSvcClient, ds.pipeline, wmFetchers, rater)
```
Comment (whynowy) [inline pkg/daemon/server/daemon_server.go:141]: There's already one in `newGRPCServer()`, move it there?

File: pkg/daemon/server/service/healthStatus.go [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+	}
+}
+
+// DefaultDataHealthResponse is the default response returned by the data health check API
+var DefaultDataHealthResponse = NewDataHealthResponse(PipelineStatusOK,
+	"Pipeline dataflow is healthy",
+	"D1")
+
+var (
+	currentPipelineStatus = DefaultDataHealthResponse
```
[the remark this answers is not in the document]
Comment (kohlisid) [reply pkg/daemon/server/service/healthStatus.go:98]: Refactored the code to include healthChecker inside Pipeline metadata query, and abstracting the implementation of ListBuffers

File: pkg/daemon/server/service/healthStatus.go [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+	// WarningState is the state of a vertex when its health is warning
+	WarningState = "warning"
+	// HealthyState is the state of a vertex when its health is healthy
+	HealthyState = "healthy"
+)
+
+const (
+	// EnableCriticalLookBack is used to enable the look back for critical state
+	EnableCriticalLookBack  = true
+	DisableCriticalLookBack = false
```
Comment (vigith) [inline pkg/daemon/server/service/healthStatus.go:62]: isn't !EnableCriticalLookBack == DisableCriticalLookBack?

File: pkg/daemon/server/service/pipeline_metrics_query_test.go
Comment (vigith) [inline pkg/daemon/server/service/pipeline_metrics_query_test.go:1]: what tests are we going add?

File: pkg/shared/ewma/interface.go [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package ewma
+
+type EWMA interface {
```
Comment (vigith) [inline pkg/shared/ewma/interface.go:3]:
```suggestion
// EWMA is exponential weighted mean average.
type EWMA interface {
```

File: pkg/shared/ewma/simple_ewma.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
+package ewma
+
+const (
+	averageAge = 30.0
+	// constDecayFactor is the default decay factor
+	constDecayFactor = 2.0 / (averageAge + 1.0)
```
Comment (vigith) [inline pkg/shared/ewma/simple_ewma.go:6]:
```suggestion
	defaultAlpha = 30.0
	// constDecayFactor is the default decay factor
	constDecayFactor = 2.0 / (defaultAlpha + 1.0)
```

File: pkg/shared/health-status-code/code_map.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
+package health_status_code
```
Comment (vigith) [inline pkg/shared/health-status-code/code_map.go:1]:
```suggestion
// package health_status_code .....
package health_status_code
```

File: server/apis/v1/health.go [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+// 5. Killed: The pipeline is killed
+// To check for vertex level status we need to check for two things,
+// 1) the number of replicas running in the vertex
+// are equal to the number of desired replicas and the pods are in running state
+// 2) If all the containers in the pod are in running state
+// if any of the above conditions are not met, the vertex is unhealthy
+func checkVertexLevelHealth(ctx context.Context, h *handler, ns string,
+	pipeline string) (*resourceHealthResponse, error) {
+	log := logging.FromContext(ctx)
+	// check if the pipeline is paused, if so, return paused status
```
Comment (vigith) [inline server/apis/v1/health.go:87]: fix

File: server/apis/v1/health.go [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+		return &resourceHealthResponse{
+			Status:  PipelineStatusInactive,
+			Message: fmt.Sprintf("pipeline %q is paused", pipeline),
+			Code:    "V7",
+		}, nil
+	}
+
+	// if the pipeline is killed, return killed status
+	// this cannot be checked at individual vertex level, hence needs to be checked here
+	// TODO(Health): Check if this is correct?
```
Comment (vigith) [inline server/apis/v1/health.go:110]: sounds correct to me.

File: pkg/daemon/server/service/healthStatus.go [hunk 10]
```diff
@@ -0,0 +1,456 @@
...
+		ticker := time.NewTicker(HealthTimeStep)
+		defer ticker.Stop()
+		for {
+			select {
+			// If the ticker ticks, check and update the health status of the pipeline.
+			case <-ticker.C:
+				// Get the current health status of the pipeline.
+				criticality, err := hc.getPipelineDataCriticality()
+				if err != nil {
+					return
```
Comment (whynowy) [inline pkg/daemon/server/service/healthStatus.go:173]: Return means quit the goroutine?

File: pkg/shared/ewma/interface.go (hunk 5 above)
Comment (whynowy) [inline pkg/shared/ewma/interface.go:3]: docs.

File: pkg/daemon/server/service/healthStatus.go [hunk 11]
```diff
@@ -0,0 +1,456 @@
...
+	// Set the health status.
+	hc.currentDataStatus = status
+}
+
+// StartHealthCheck starts the health check for the pipeline.
+// The ticks are generated at the interval of HealthTimeStep.
+func (hc *HealthChecker) StartHealthCheck(ctx context.Context) {
+	// Goroutine to listen for ticks
+	// At every tick, check and update the health status of the pipeline.
+	go func(ctx context.Context) {
```
Comment (whynowy) [inline pkg/daemon/server/service/healthStatus.go:162]: Shouldn't the function `StartHealthCheck` be blocked? It's using a goroutine to start it in the daemon server.

File: server/apis/v1/health.go [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+	log := logging.FromContext(ctx)
+
+	// create a cache key for the pipeline
+	// It is a combination of namespace and pipeline name
+	// In the form of <namespace>-<pipeline>
+	cacheKey := fmt.Sprintf("%s-%s", ns, pipeline)
+
+	// check if the pipeline status is cached
+	if status, ok := hc.resourceStatusCache.Get(cacheKey); ok {
+		log.Info("pipeline status from cache: ", status)
```
Comment (whynowy) [inline server/apis/v1/health.go:58]: Minor: Capital `P` in `pipeline`

File: pkg/daemon/server/service/healthStatus.go (hunk 10 above)
Comment (yhl25) [inline pkg/daemon/server/service/healthStatus.go:173]: log the error

**Review by yhl25 — COMMENTED** [review]
I see a lot of methods and types exposed in `healthStatus.go`, avoid exposing types and methods if they are not used in other packages

File: pkg/daemon/server/service/healthStatus.go [hunk 13]
```diff
@@ -0,0 +1,456 @@
...
+	"github.com/numaproj/numaflow/pkg/shared/ewma"
+	sharedqueue "github.com/numaproj/numaflow/pkg/shared/queue"
+)
+
+const (
+	// HealthSlidingWindow is the total time window to compute the health of a vertex
+	HealthSlidingWindow = 5 * time.Minute
+
+	// HealthTimeStep is the frequency at which the health of a vertex is computed
+	HealthTimeStep = 10 * time.Second
```
[the remark this answers is not in the document]
Comment (kohlisid) [reply pkg/daemon/server/service/healthStatus.go:22]: This should be running independently! So we should be fine

File: pkg/daemon/server/service/healthStatus.go (hunk 11 above)
Comment (kohlisid) [reply pkg/daemon/server/service/healthStatus.go:162]: Updated. Please check if it seems fine now

File: pkg/daemon/server/service/healthStatus.go (hunk 10 above)
Comment (kohlisid) [reply pkg/daemon/server/service/healthStatus.go:173]: Returning unknown status now

**Review by vigith — APPROVED** [review]

**Comment by whynowy** [comment]
@kohlisid - resolve the conflicts?

**Review by whynowy — APPROVED** [review]

**Pull request merged**